pennylane/dla/lie_closure.py

```diff
@@ -24,6 +26,132 @@
 from pennylane.pauli import PauliSentence, PauliWord
 
 
+def lie_closure(
+    generators: Iterable[Union[PauliWord, PauliSentence, Operator]],
+    max_iterations: int = 10000,
+    verbose: int = 0,
+    pauli: bool = False,
+) -> Iterable[Union[PauliWord, PauliSentence, Operator]]:
+    r"""Compute the Lie closure over commutation of a set of generators
+
+    The Lie closure, pronounced "Lee" closure, is a way to compute the so-called dynamical Lie algebra (DLA) of a set of generators :math:`\mathcal{G} = \{G_1, .. , G_N\}`.
+    For such generators, one computes all nested commutators :math:`[G_i, [G_j, .., [G_k, G_\ell]]]` until no new operators are generated from commutation.
+    All these operators together form the DLA, see e.g. section IIB of `arXiv:2308.01432 <https://arxiv.org/abs/2308.01432>`__.
+
+    Args:
+        generators (Iterable[Union[PauliWord, PauliSentence, Operator]]): generating set for which to compute the
+            Lie closure.
+        max_iterations (int): maximum depth of nested commutators to consider. Default is ``10000``.
+        verbose (Union[int, bool]): verbosity during Lie closure calculation. Default is ``0``.
+        pauli (bool): Indicates whether it is assumed that :class:`~PauliSentence` instances are input and returned.
+            This can help with performance to avoid unnecessary conversions from :class:`~PauliSentence` to :class:`~Operator`
+            or :class:`~PauliWord` and vice versa. Note that the input in that case also has to be a list of :class:`~PauliSentence` instances.
```

Comment:
If you don't pass `PauliSentence`, do we check for this or we might just expect an obscure error?

Reply:
Yes, it would indeed be an obscure failure at the moment.

The point of this option is performance and thus to avoid checks

```pycon
>>> ops = [X(0), X(1), Y(0) @ Y(1)]
>>> qml.dla.lie_closure(ops, pauli=True)
AttributeError: 'PauliX' object has no attribute 'keys'
```

Reply:
That being said I think a quick type check is worth it. Added it, let me know if you want me to remove it again